File: C/27_ShortestDistanceBacktracking/BOJ_09_11780_플로이드2.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#define math_min(a, b) a < b ? a : b
int** idxArr;
const int INF = INT_MAX / 2;
void floydWarshall(int** route, int n) {
	for (int k = 1; k <= n; k++) {
		for (int i = 1; i <= n; i++) {
			for (int j = 1; j <= n; j++) {
				if (route[i][j] > route[i][k] + route[k][j]) {
					route[i][j] = route[i][k] + route[k][j];
					idxArr[i][j] = idxArr[k][j];
				}
			}
		}
	}
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			printf("%d ", route[i][j] != INF ? route[i][j] : 0);
		}
		printf("\n");
	}
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			if (!idxArr[i][j]) {
				printf("0\n");
			}
			else {
				int* stack = (int*)malloc(sizeof(int) * n);
				int idx = -1, k = idxArr[i][j];
				stack[++idx] = j;
				while (i != k) {
					stack[++idx] = k;
					k = idxArr[i][k];
				}
				stack[++idx] = i;
				printf("%d ", idx + 1);
				while (idx >= 0) {
					printf("%d ", stack[idx--]);
				}
				printf("\n");
				free(stack);
			}
		}
	}
}
main() {
	int n, m;
	scanf("%d %d", &n, &m);
	int** route = (int**)malloc(sizeof(int*) * (n + 1));
	idxArr = (int**)malloc(sizeof(int*) * (n + 1));
	for (int i = 1; i <= n; i++) {
		route[i] = (int*)malloc(sizeof(int) * (n + 1));
		idxArr[i] = (int*)malloc(sizeof(int) * (n + 1));
		for (int j = 1; j <= n; j++) {
			route[i][j] = INF;
			idxArr[i][j] = 0;
		}
		route[i][i] = 0;
	}
	int a, b, c;
	while (m--) {
		scanf("%d %d %d", &a, &b, &c);
		route[a][b] = math_min(route[a][b], c);
		idxArr[a][b] = a;
	}
	floydWarshall(route, n);
}

File: JAVA/27_ShortestDistanceBacktracking/BOJ_09_11780_플로이드2.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {

    final static int INF = Integer.MAX_VALUE / 2;
    static Integer[][] idxArr;

    public static String floydWarshall(int[][] route, int n) {
        StringBuilder sb = new StringBuilder();
        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    if (route[i][j] > route[i][k] + route[k][j]) {
                        route[i][j] = route[i][k] + route[k][j];
                        idxArr[i][j] = idxArr[k][j];
                    }
                }
            }
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                sb.append(route[i][j] != INF ? route[i][j] : 0).append(" ");
            }
            sb.append("\n");
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (idxArr[i][j] == null) {
                    sb.append(0);
                } else {
                    Stack<Integer> stack = new Stack<>();
                    stack.push(j);
                    int k = idxArr[i][j];
                    while (i != k) {
                        stack.push(k);
                        k = idxArr[i][k];
                    }
                    stack.push(i);
                    sb.append(stack.size()).append(" ");
                    while (!stack.isEmpty()) {
                        sb.append(stack.pop()).append(" ");
                    }
                }
                sb.append("\n");
            }
        }
        return sb.toString();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());
        int[][] route = new int[n + 1][n + 1];
        idxArr = new Integer[n + 1][n + 1];
        for (int i = 1; i <= n; i++) {
            Arrays.fill(route[i], INF);
            route[i][i] = 0;
        }
        while (m-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            route[a][b] = Math.min(route[a][b], c);
            idxArr[a][b] = a;
        }
        System.out.print(floydWarshall(route, n));
    }
}